### PR TITLE
For #12118 : Set default screen orientation if playing media in pip

### PR DESCRIPTION
**Pull request by indurs**
This prevents the activity being locked in landscape mode while it's opened for new intents when the pip mode is active

Co-Authored-By: Mugurell <Mugurell@users.noreply.github.com>

When the search widget is clicked

https://user-images.githubusercontent.com/12825812/168533587-871f0354-0dda-4193-8da1-ffb51f809a1e.mp4

From shortcut 

https://user-images.githubusercontent.com/12825812/168533629-20336e3f-1292-4c67-a899-9308e4dbda38.mp4

When a page is shared to Fenix

https://user-images.githubusercontent.com/12825812/168533674-5b0f236c-582c-4a0c-a1a1-fc1392244bf8.mp4

When a link is shared to Fenix

https://user-images.githubusercontent.com/12825812/168534175-c492049c-b3be-44ad-8d54-d0541b850b36.mp4

@Mugurell , 
Following changes have been made to MediaSessionFullscreenFeature

Activity's screen orientation is set to ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED while in pip mode.
Also added a condition to check whether the tab is an active media session tab before setting the orientation. 
In some scenarios the flow in MediaSessionFullscreenFeature filters the tab that has fullscreeen enabled in mediasessionState even after a  page is loaded(For instance when a page is shared from Focus). And this results in processFullscreen() being called and the activity is set to landscape. So the condition will check whether the selected tab is an active media tab before setting the orientation. 

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by Mugurell** [comment]
"In some scenarios the flow in MediaSessionFullscreenFeature filters the tab that has fullscreeen enabled in mediasessionState even after a page is loaded(For instance when a page is shared from Focus). And this results in processFullscreen() being called and the activity is set to landscape. So the condition will check whether the selected tab is an active media tab before setting the orientation."

This sounds like another bug?
Can you share some steps to reproduce?
Not sure about this `tab that has fullscreeen enabled in mediasessionState even after a page is loaded(For instance when a page is shared from Focus)` scenario.

cc @rocketsroger

**Comment by Mugurell** [comment]
Seems like there is one test failing that seems related to these changes - https://firefoxci.taskcluster-artifacts.net/Mx6BWBu-R_utYTr9I6KMpg/0/public/reports/tests/testReleaseUnitTest/index.html
Please try and fix it (more to test the new logic would be also great), here to help if needed.

**Comment by Mugurell** [comment]
Thank you for investigating all the scenarios!
The results look great!
Also curious about @rocketsroger's opinion of these changes.

**Comment by indurs** [comment]
Will upload videos regarding the bug.

**Comment by indurs** [comment]
When we come from shortcut while the activity is in PIP mode

While the shortcut page is loading, the flow in MediaSessionFullscreenFeature filters out the tab that has 
``` mediaSessionState != null && mediaSessionState!!.fullscreen```. This results in processFullscreen() method being invoked and the activity's orientation is set to landscape.So the loading page (the activity) goes to landscape mode. And when the mediaSessionState becomes null processFullscreen() is called and the activity's orientation is set to portrait. 


https://user-images.githubusercontent.com/12825812/168743324-35ff5b67-b6d3-4351-a0e0-0e52abfba7e1.mp4


When a page is shared from Focus to Fenix

When the page is loading the flow in MediaSessionFullscreenFeature filters out the tab that has 
``` mediaSessionState != null && mediaSessionState!!.fullscreen``` 
 So the loading page (the activity) goes to landscape mode. The activity is locked to landscape mode since the mediaSessionState is not set to null and the mediaSessionState.fullscreen is not set to false till the end. 


https://user-images.githubusercontent.com/12825812/168743444-413ae037-3adc-4bfe-b49e-ef936b64764d.mp4


This condition 
```if (store.state.selectedTabId == activeState.id) ``` sets the activity's orientation only if the tab is an active media tab.

1. For shortcut, while loading the page the media session is removed and hence the mediasessionstate becomes null.
2. For sharing page or link from Focus, the media session is not removed till the end. So the media tab will have the mediasessionState and its fullscreen remains true even after the page is loaded.

Do we need to call components.useCases.sessionUseCases.exitFullscreen method, so that the mediaSessionState is cleared? Probably we can call that method before any intentProcessors is called in handleNewIntent() of HomeActivity. It can run in background.
```
if(components.core.store.state.findActiveMediaTab() !=null){
            lifecycleScope.launch(IO) {
                var tab = components.core.store.state.findActiveMediaTab()
                components.useCases.sessionUseCases.exitFullscreen(tab!!.id)
            }
        }
}
```
Also I'm not sure how the flow works with ifChanged() method in MediaSessionFullscreenFeature. I think processFullscreen() is invoked  whenever there is any change in mediaSessionState not just when the fullscreen value is changed.

**Comment by Mugurell** [comment]
> Reason for adding the condition `if (store.state.selectedTabId == activeState.id)`
> ...

Thank you for the explanation and investigating the entire flow to this extent!
This proposed change makes sense to me and I do appreciate the benefits seen while comparing the videos.

> Do we need to call components.useCases.sessionUseCases.exitFullscreen method, so that the mediaSessionState is cleared? Probably we can call that method before any intentProcessors is called in handleNewIntent() of HomeActivity. It can run in background.

If we don't expressly need this now it would probably be better to focus on fewer moving parts.

> 
> Also I'm not sure how the flow works with ifChanged() method in MediaSessionFullscreenFeature. I think processFullscreen() is invoked whenever there is any change in mediaSessionState not just when the fullscreen value is changed.

Seems like it to me also.
The current check from `processFullscreen` for the first tab with a media state changed being in fullscreen and then your proposed check to ensure that that is also the currently selected tab seem to ensure we'll lock orientations only for when the currently selected tab is in fullscreen and unlock otherwise which sounds great.

**Comment by indurs** [comment]
Thank You @Mugurell !! We can look into the flow later. But now we can add the components.useCases.sessionUseCases.exitFullscreen method in HomeActivity right? and then update the MediaSessionFullscreenFeatureTest.

**Comment by Mugurell** [comment]
> Thank You @Mugurell !! We can look into the flow later. But now we can add the components.useCases.sessionUseCases.exitFullscreen method in HomeActivity right? and then update the MediaSessionFullscreenFeatureTest.

The changes seem okay here and I think we can review and land this before Fenix. (nothing will break in Fenix, the fullscreen flows would be improved but the crash which started all this won't be fixed. Please correct me on this.)

You can work on having everything ready in Fenix also (by using a local version of `android-components`) but either way that will probably have to be merged after the changes here to prevent this situation - https://github.com/mozilla-mobile/fenix/pull/24951#pullrequestreview-972194155 .

**Comment by indurs** [comment]
`The changes seem okay here and I think we can review and land this before Fenix. (nothing will break in Fenix, the fullscreen flows would be improved but the crash which started all this won't be fixed. Please correct me on this.)`

You are right. The crash when we search from the search widget won't be fixed and nothing will break in Fenix. 

If we land just the changes for MediaSessionFullscreenFeature, it will fix the following issues
1. The activity 's orientation won't be locked in landscape when we share a page or link from Focus to Fenix
2. The activity 's orientation won't be rotated to landscape and  then to portrait when we navigate to Fenix from a shortcut

As you mentioned we'll first land the changes for MediaSessionFullscreenFeature in AC.

After we land the changes in AC and then in Fenix, the crash and the orientation issue related to search widget will be fixed.

`You can work on having everything ready in Fenix also (by using a local version of android-components) but either way that will probably have to be merged after the changes here to prevent this situation - https://github.com/mozilla-mobile/fenix/pull/24951#pullrequestreview-972194155 `

I'll keep everything ready in Fenix. We'll merge the changes in Fenix after we land the changes in AC

**Comment by indurs** [comment]
@Mugurell , I've added test cases for MediaSessionFullscreenFeature. Please let me know if I've missed test cases for any scenarios. 
lint-detektAndKtlint is failing.  I've arranged the imports for MediaSessionFullscreenFeatureTest. But still I get the following error message. 
Imports must be ordered in lexicographic order without any empty lines in-between with "java", "javax", "kotlin" and aliases in the end

File: components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt
Comment (Mugurell) [inline components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:39]: Here and below: If using the GIVEN-WHEN-THEN naming scheme please capitalize these words.

**Review by Mugurell — COMMENTED** [review]
This looks good to me.
The tests are a bit too complex and having smaller ones targeting one scenario at a time would help but seems all cases are covered so there are only three small things to take care of:
- optimize imports. Using `⌃⌥O` in AndroidStudio on MacOS helped me get https://gist.github.com/Mugurell/9d72ae62fbe1bb8347232b5d2ce0ead2 which has both detekt and ktlint passing.
- capitalize `then` to `THEN`
- squash all commits into one.

and then this seems ready to land.

Thank you for all your efforts on this!

File: components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt
Comment (indurs) [reply components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:39]: Thanks You for the suggestions. Will update the test cases.

Comment (indurs) [reply components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:39]: Sure. Will make the changes. Thank You!

**Comment by indurs** [comment]
@Mugurell , shall I  update the changelog to include this issue fix?

File: components/feature/media/build.gradle
Comment (Mugurell) [inline components/feature/media/build.gradle:49]: This doesn't seem to be needed?

File: components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt
Comment (Mugurell) [inline components/feature/media/src/test/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeatureTest.kt:127]: nit: Here and below: "Given" is not capitalized.

**Review by Mugurell — COMMENTED** [review]
LGTM.
Seems like just adding adding a changelog entry and removing the `feature-session` is needed.

**Comment by Mugurell** [comment]
> @Mugurell , shall I update the changelog to include this issue fix?

Yes, thank you!
Can you please also update the commit message to not contain all the duplicated titles? To have it just as it was one commit from the start?
This seems ready to land but to get the full result and be able to test it, it seems better to also wait for the Fenix changes to be ready and then merge both patches around the same time.

**Comment by indurs** [comment]
@Mugurell , I've committed the changes. Thank you for reviewing and for the suggestions.

**Review by Mugurell — COMMENTED** [review]
LGTM, thank you!

**Comment by Mugurell** [comment]
Waiting to also have the Fenix patch - https://github.com/mozilla-mobile/fenix/pull/24951 ready to merge

**Review by Mugurell — APPROVED** [review]

**Review by rocketsroger — APPROVED** [review]
Tried it out, works great! Thanks!

**Comment by Mugurell** [comment]
PR merged in https://github.com/mozilla-mobile/android-components/pull/12239.

**Pull request closed without merging**
